sys/net/link_layer/Kconfig.lorawan

```diff
@@ -0,0 +1,518 @@
+    bool "TXPower 14"
+
+config LORAMAC_DEFAULT_TX_POWER_15
+    bool "TXPower 15"
+
+endif # LORAMAC_REGION_US_915
+
+endchoice
+
+config LORAMAC_DEFAULT_TX_PORT
+    int "MAC TX port"
+    default 2
+    range 1 223
+    help
+        Configure the default MAC TX port. The TX port denotes the port field
+        `FPort` in the payload and is commonly used on the destination
+        side (Application Server) to distinguish different payload formats. For
+        example it can be used to identify different sensor types/values from
+        the same end-node and hence different payload formats can be assigned to
+        the different values based on value of port.
+
+choice
+    bool "MAC TX mode"
+    default LORAMAC_DEFAULT_TX_MODE_CNF
+    help
+        Configure the type of uplink packet (confirmed/unconfirmed). A confirmed
+        uplink will expect an acknowledgment from the network server; otherwise,
+        the message will be retransmitted by the number indicated by
+        CONFIG_LORAMAC_DEFAULT_RETX, whereas an unconfirmed message will not
+        expect any acknowledgment back from the network server.
+
+config LORAMAC_DEFAULT_TX_MODE_CNF
+    bool "Confirmable"
+
+config LORAMAC_DEFAULT_TX_MODE_UNCNF
+    bool "Unconfirmable"
+
+endchoice
+
+config LORAMAC_ADVANCED_OPTIONS
+    bool "Enable advanced options"
+
+config LORAMAC_DEFAULT_RETX
+    int "Uplink retransmissions" if LORAMAC_ADVANCED_OPTIONS
+    depends on USEMODULE_GNRC_LORAWAN || USEMODULE_RN2XX3
+    default 5
+    range 0 255
+    help
+        Configure the maximum number of retransmissions to be used for a
+        confirmed uplink packet, if no downlink acknowledgment is received from
+        the network.
+
+config LORAMAC_DEFAULT_ADR
+    bool "Enable ADR" if LORAMAC_ADVANCED_OPTIONS
+    depends on USEPKG_SEMTECH_LORAMAC || USEMODULE_RN2XX3
+    help
+        Enable or disable Adaptive Data Rate (ADR). If enabled the end node will
+        inform the network server about the status of ADR using the ADR field in
+        uplink data packet. The network server will then optimize the data rate
+        and the transmission power of the end node based on the information
+        collected from the network.
+
+config LORAMAC_DEFAULT_LINKCHK
+    int "Link check interval (in seconds)" if LORAMAC_ADVANCED_OPTIONS
+    depends on USEMODULE_RN2XX3
+    default 0
+    help
+        Configure the default link check interval (in seconds). 0 means the link
+        check process is disabled. The link check process is used by an
+        end-device to validate its connectivity with the network. The frame has
+        no payload. When a `LinkCheckReq` is received by the network server via
+        one or multiple gateways, it responds with a `LinkCheckAns` MAC command.
+
+config LORAMAC_DEFAULT_RX1_DELAY
+    int "RX1 window delay (in milliseconds)" if LORAMAC_ADVANCED_OPTIONS
+    depends on USEMODULE_GNRC_LORAWAN || USEMODULE_RN2XX3
+    default 1000
```

Comment:
Is there a range that could be added?

Reply:
As per LoRaWAN Spec, this is supposed to be 1s. When it comes to range, the rationale is hard to define. Say for a 64byte  packet size (usually the maximum packet size in public nw's) in DR0(SF12/BW125) in EU868, the airtime will be ~9s. However, a packet with a theoretical maximum of 238 bytes needs  ~853 seconds airtime. Hence I don't think range makes sense here.

Reply:
This delay is independent of the Time on Air. This is the delay between the end of the uplink and the first RX window. In practice, we can this value is 4 bit long. In worst case, 8 bit (range from 1-255).
The regional parameter limit the Time on Air to ~2.6 seconds (ie.g it's not possible to send more than 51 bytes in DR0). For 51 bytes fives a ToA of ~2.6 seconds, considering the symbol time is 32 ms.

Reply:
Please correct me if I am wrong, isn't the time calculated based on the time for the uplink to reach the gateway + network latency. The time for uplink to reach the gateway is what I meant by time on air.  I used this tool < https://avbentem.github.io/airtime-calculator/ttn/eu868/52,12 > to calculate the TOA, assuming an overhead size of 15 bytes in each packet.

Reply:
No, this is the time between the last bit of the uplink packet (at the gateway) and the moment the node opens the first RX window. Network latency is not considered, since the gateway timestamps the uplink frames.

This is simply used to know at what time the RX window has to be open. For the node, this is the time between the TX_DONE event and the SPI command that puts the radio in LISTEN.

> https://avbentem.github.io/airtime-calculator/ttn/eu868/52,12 > to calculate the TOA, assuming an overhead size of 15 bytes in each packet.

It gives me 2,957.3 seconds for that. Did you check the "airtime"?

Reply:
I checked the wrong section, which says about FUP.

Reply:
> Edit: I think I got confused with Air time and ToA. ToA will be in milliseconds range, hence may not be really relevant here, I suppose. OTAA has a max RX1 delay of 5 seconds. So range 1 5 should work, right ?

In practice I don't think a NS could configre more than 5 seconds. So I think this should be fine. But note this parameter is still in milliseconds. Therefore the range should be `range 1000 5000`